### PR TITLE
vimc-4137: test as many functions as possible, and fix bugs along testing

### PR DESCRIPTION
**Pull request by xiangli313**
Task:
As title.

Youtrack:
https://mrc-ide.myjetbrains.com/youtrack/issue/VIMC-4137

Hi, Rob. May be a stupid question. I have tested a lot of functions. But the test coverage is still low (<30%). Is it that I am testing in a wrong way?

### CODE REVIEW

File: R/impact_central.R [hunk 1]
```diff
@@ -24,27 +24,32 @@ get_raw_impact_details <- function(con, meta1, burden_outcome, is_under5 = FALSE
   v <- determine_vaccine_delivery(meta1)
   i <- unique(grepl("routine", v))
   j <- unique(grepl("campaign", v))
-  if(i & j & meta1$method[1] == "method2a"){
+  if(any(i) & any(j) & meta1$method[1] == "method2a"){
```
Comment (r-ash) [inline R/impact_central.R]: Generally R docs suggest preferring long form of boolean operators i.e. `&&` and `||` over `&` and `|`.  From docs

> The longer form is appropriate for programming control-flow and typically preferred in if clauses.

File: R/impact_central.R [hunk 2]
```diff
@@ -24,27 +24,32 @@ get_raw_impact_details <- function(con, meta1, burden_outcome, is_under5 = FALSE
...
+    country_constrain <- ""
+  } else{
+    countries_to_extract <- DBI::dbGetQuery(con, sprintf("SELECT nid FROM country WHERE id IN %s",
+                                                         sql_in(countries_to_extract)))
+    country_constrain <- sprintf("AND country IN %s", sql_in(countries_to_extract, text_item = FALSE))
+  }
 
   # set up db extraction sql queries
-  if((meta1$method[1] == "method0") | (meta1$method[1] == "method2a" & j)){
+  if((meta1$method[1] == "method0") | (meta1$method[1] == "method2a" & any(j))){
```
Comment (r-ash) [inline R/impact_central.R]: Same with this one here - I think should use `||` and `&&`

File: R/impact_central.R [hunk 3]
```diff
@@ -140,27 +144,30 @@ impact_by_year_of_vaccination <- function(meta1, raw_impact, fvps, fvps_updates
...
       raw_impact <- raw_impact[raw_impact$time >= min(vaccination_years), ]
     }
     tot_impact <- stats::aggregate(value ~ country + burden_outcome, raw_impact, sum, na.rm = TRUE)
     tot_fvps <- stats::aggregate(fvps~country, fvps, sum, na.rm = TRUE)
     d <- merge_by_common_cols(tot_impact, tot_fvps, all = TRUE)
   } else if (method == "method2b"){
-    cohort_impact <- raw_impact[raw_impact$time %in% cohorts_vaccinated, c("country", "time", "burden_outcome", "value")]
-    cohort_fvps <- stats::aggregate(fvps ~ country + time, fvps, sum, na.rm = TRUE)
+    cohort_impact <- raw_impact[raw_impact$time %in% (min(fvps$time):max(fvps$time)), c("country", "time", "burden_outcome", "value")]
+    cohort_fvps <- aggregate(fvps ~ country + time, fvps, sum, na.rm = TRUE)
```
Comment (r-ash) [inline R/impact_central.R]: Removing the `stats::` part will trigger a note or warning in R CMD check - it's good to try and remove these just to reduce noise. Unless we want to use aggregate from one of the other packages - in that case we should include the correct package.

**Review by r-ash — CHANGES_REQUESTED** [review]
I think the coverage issues will be because the tests added are being skipped on the test system. It doesn't have `MONTAGU_DB_PASSWORD` and so those tests will be skipped during the coverage. It also means that those tests won't be being run on the build server so we need to make sure they run locally.

I think switching to using the test postgres db which I added for the stochastic helper function might make it a bit easier - we would just have to create some dummy tables as part of the test setup. That can be done in another ticket though - and maybe let's talk about it at some point.

Indeed running locally I get an error in the tests
```
test-impact.R:89: error: test if vimpact functions are working as expected for central estimates
replacement has 1 row, data has 0
Backtrace:
 1. base::lapply(...) tests/testthat/test-impact.R:89:2
 2. vimpact:::FUN(X[[i]], ...)
 3. vimpact::get_raw_impact_details(con = con_test, meta1, burden_outcome = "deaths") tests/testthat/test-impact.R:89:22
 5. base::`$<-.data.frame`(`*tmp*`, "index", value = 6L) R/impact_central.R:82:2
```

Can you recreate that?

File: R/impact_central.R (hunk 1 above)
Comment (xiangli313) [reply R/impact_central.R]: Cool. This is done.

File: R/impact_central.R (hunk 2 above)
Comment (xiangli313) [reply R/impact_central.R]: This is done.

File: R/impact_central.R (hunk 3 above)
Comment (xiangli313) [reply R/impact_central.R]: I see. I will add stats:: back to avoid warnings.

**Comment by xiangli313** [comment]
Hi @rob. Thanks for your helpful comments. I was doing tests for Measles and YF. When the two diseases passes tests, I added in HepB without testing. And the error comes from HepB. The reason is HepB IC-Hallett model dose not model PINE countries (I was only testing for four countries) for all their scenarios. This results in 0 rows data for some scenarios. I will try to resolve it.

**Comment by rob** [comment]
This is definitely beyond my pay grade!

**Comment by xiangli313** [comment]
Hi @r-ash , now the error should have gone. Please check from your side. We talked about what to test for impact calculations. Our thoughts are 1. creating dummy data can be complex, as we will need to mock up a lot of db tables to make things running; 2. we want to test vimpact results against existing estimates we used to share with funders, so that we are confident that we didn't dump them rubbish, and vimpact will not neither. So, I guess having access to Montagu for testing is the right approach at least for now. Although travis will skip those tests, we know they are working internally.

**Review by r-ash — CHANGES_REQUESTED** [review]
Xiang, can confirm this is working locally now. Can you do one more thing and add some instructions to the README for how you setup to make the tests work locally? Possibly adding a script if sensible - something like the notes for using postgres https://github.com/vimc/vimpact/blob/master/README.md#running-tests-which-rely-on-postgresql-database

Rich and I can then use those instructions to setup the CI system to run these tests.

**Review by r-ash — APPROVED** [review]

**Pull request merged**